lib/fileset.nix

```
@@ -0,0 +1,993 @@
+  However this function can do the same thing incrementally, so it can already start printing the result as the first lines are known.
+
+  The `expand` argument (false by default) controls whether all files should be printed individually.
+
+  Type:
+    trace :: { expand :: Bool ? false } -> FileSet -> Any -> Any
+
+  Example:
+    trace {} (unions [ ./foo.nix ./bar/baz.c ./qux ])
+    =>
+    trace: /home/user/src/myProject
+    trace: - bar
+    trace:   - baz.c (regular)
+    trace: - foo.nix (regular)
+    trace: - qux (recursive directory)
+    null
+
+    trace { expand = true; } (unions [ ./foo.nix ./bar/baz.c ./qux ])
+    =>
+    trace: /home/user/src/myProject
+    trace: - bar
+    trace:   - baz.c (regular)
+    trace: - foo.nix (regular)
+    trace: - qux
+    trace:   - florp.c (regular)
+    trace:   - florp.h (regular)
+    null
+  */
+  trace = { expand ? false }: maybeFileset:
+    let
+      fileset = _coerce "trace" "second argument" maybeFileset;
+      simpleTree = _simplifyTree expand fileset._base fileset._tree;
+    in
+    _prettyFoldl' (acc: el: builtins.trace el acc) (x: x)
+      fileset._base
+      simpleTree;
+
+  /*
+  The same as `lib.fileset.trace`, but instead of taking an argument for the value to return, the given file set is returned instead.
+
+  Type:
+    traceVal :: { expand :: Bool ? false } -> FileSet -> FileSet
+  */
+  traceVal = { expand ? false }: maybeFileset:
+    let
+      fileset = _coerce "traceVal" "second argument" maybeFileset;
+      simpleTree = _simplifyTree expand fileset._base fileset._tree;
+    in
+    _prettyFoldl' (acc: el: builtins.trace el acc) fileset
+      fileset._base
+      simpleTree;
+
+  /*
+  The same as `lib.fileset.trace`, but instead of tracing each line, the result is returned as a string.
+
+  Type:
+    pretty :: { expand :: Bool ? false } -> FileSet -> String
+  */
+  pretty = { expand ? false }: maybeFileset:
+    let
+      fileset = _coerce "pretty" "second argument" maybeFileset;
+      simpleTree = _simplifyTree expand fileset._base fileset._tree;
+    in
+    _prettyFoldl' (acc: el: "${acc}\n${el}") ""
+      fileset._base
+      simpleTree;
+
+  /*
+  The file set containing all files that are in either of two given file sets.
+  Recursively, the first argument is evaluated first, only evaluating the second argument if necessary.
+
+      union a b = a ⋃ b
+
+  Type:
+    union :: FileSet -> FileSet -> FileSet
+  */
+  union = lhs: rhs:
+    let
+      normalised = _normaliseBase "union" [
+        {
+          context = "first argument";
+          value = lhs;
+        }
+        {
+          context = "second argument";
+          value = rhs;
+        }
+      ];
+    in
+    _create normalised.commonBase
+      (_unionTree
+        (elemAt normalised.trees 0)
+        (elemAt normalised.trees 1)
+      );
+
+  /*
+  The file containing all files from that are in any of the given file sets.
```

```suggestion
  The set containing all files that are in any of the given sets.
```

Also in the other comments I think we don't have to say "file set" every time. Such repetitiveness is quite a strain to read.